### ISSUE TITLE
关于固定随机种子，但SeCo与ToCo的每次实验结果都不同的问题

### ISSUE THREAD

**Issue by Jingfeng-Tang** [zwyang6/SeCo#7]
同学你好，感谢您分享代码！
我跑了两次SeCo，但是实验结果却不同。在第1个iter到第2000个iter的损失函数值都是相同的，但是第2000个iter后，两次实验打印的各个损失值都不相同了。（我注意到您的代码已经固定随机种子了，我的这两次实验也并未修改任何超参。）
我在固定随机种子的代码后打印随机的tensor，两次实验的tensor元素值是相同的。
我在另一台服务器上重新跑了两次实验，第2000个iter后的损失值也不同。
我也跑了两次ToCo的代码，发现也出现了这个问题。
`# warmup
    if n_iter <= 1000:
        loss = 1.0 * cls_loss + 1.0 * cls_loss_aux + args.w_ptc * ptc_loss + 0.0 * seg_loss + 0.0 * reg_loss
    elif n_iter <= 2000:
        loss = 1.0 * cls_loss + 1.0 * cls_loss_aux + args.w_ptc * ptc_loss + 0.0 * seg_loss + 0.0 * reg_loss
    else:
        loss = 1.0 * cls_loss + 1.0 * cls_loss_aux + args.w_ptc * ptc_loss + args.w_seg * seg_loss + args.w_reg * reg_loss`
在2000个iter后，损失函数中加入了seg_loss和reg_loss，我不知道是不是这两个损失函数造成的，还请您帮助解答疑惑，谢谢。

**Comment by zwyang6** [comment]
您好，我也注意到类似问题，很多单阶段的wsss均有此现象。我自己排查得到的原因是2k个iter后分割部分的引入所导致的。具体的原因可能是您所说的两个loss所导致，更大可能是cam取阈值所生成的伪标签每次有所浮动变化。


这个问题我暂时是每次训练均保存好权重以保证结果的可复现。希望能对你有所帮助。

--------------原始邮件--------------
发件人："Jingfeng&nbsp;Tang ***@***.***&gt;;
发送时间：2024年8月29日(星期四) 晚上9:03
收件人："zwyang6/SeCo" ***@***.***&gt;;
抄送："Subscribed ***@***.***&gt;;
主题：[zwyang6/SeCo] 关于固定随机种子，但SeCo与ToCo的每次实验结果都不同的问题 (Issue #7)
-----------------------------------

 

 
同学你好，感谢您分享代码！
 我跑了两次SeCo，但是实验结果却不同。在第1个iter到第2000个iter的损失函数值都是相同的，但是第2000个iter后，两次实验打印的各个损失值都不相同了。（我注意到您的代码已经固定随机种子了，我的这两次实验也并未修改任何超参。）
 我在固定随机种子的代码后打印随机的tensor，两次实验的tensor元素值是相同的。
 我在另一台服务器上重新跑了两次实验，第2000个iter后的损失值也不同。
 我也跑了两次ToCo的代码，发现也出现了这个问题。
 # warmup if n_iter <= 1000: loss = 1.0 * cls_loss + 1.0 * cls_loss_aux + args.w_ptc * ptc_loss + 0.0 * seg_loss + 0.0 * reg_loss elif n_iter <= 2000: loss = 1.0 * cls_loss + 1.0 * cls_loss_aux + args.w_ptc * ptc_loss + 0.0 * seg_loss + 0.0 * reg_loss else: loss = 1.0 * cls_loss + 1.0 * cls_loss_aux + args.w_ptc * ptc_loss + args.w_seg * seg_loss + args.w_reg * reg_loss
 在2000个iter后，损失函数中加入了seg_loss和reg_loss，我不知道是不是这两个损失函数造成的，还请您帮助解答疑惑，谢谢。

**Comment by Jingfeng-Tang** [comment]
我又跑了几次实验，互斥地加上seg_loss和reg_loss，发现均出现损失值不一致的现象。
我发现这两个损失都用了`refined_pseudo_label`，我将cam中的某个数值输出观察两次实验数值是否不同，然后一直定位到以下代码。
在模型生成cam时就已经出现不一致了。
`def multi_scale_cam2(model, inputs, scales):
    '''process cam and aux-cam'''
    # cam_list, tscam_list = [], []
    b, c, h, w = inputs.shape
    with torch.no_grad():
        inputs_cat = torch.cat([inputs, inputs.flip(-1)], dim=0)

        _cam_aux, _cam = model(inputs_cat, cam_only=True)
        print(f'_cam: {_cam[0,0,1,1]}')`

**Comment by Jingfeng-Tang** [comment]
还是谢谢您分享了代码，后面我也还是保存权重文件训练吧。谢谢。

**Comment by wxqlab** [comment]
@Jingfeng-Tang 您好，我现在训练也发现了这个问题，请问您解决这个问题了嘛？我发现每次训练的实验结果波动非常大，直接掉点掉到个位数的miou，您训练的时候是否出现这么大的性能差异吗？

**Comment by Jingfeng-Tang** [comment]
没有解决，但是好像性能没有掉到个位数的miou。

**Comment by wxqlab** [comment]
@Jingfeng-Tang 您测试出来的性能大概会有多少个点的波动呀？

**Comment by Jingfeng-Tang** [comment]
> [@Jingfeng-Tang](https://github.com/Jingfeng-Tang) 您测试出来的性能大概会有多少个点的波动呀？

抱歉我已经毕业，看不到服务器的日志，我记得大约是3、4个点的波动吧。

**Comment by zwyang6** [comment]
> [@Jingfeng-Tang](https://github.com/Jingfeng-Tang) 您好，我现在训练也发现了这个问题，请问您解决这个问题了嘛？我发现每次训练的实验结果波动非常大，直接掉点掉到个位数的miou，您训练的时候是否出现这么大的性能差异吗？

您好 感谢关注seco！ 由于毕业以及换方向等原因，我已经近一年没有进行WSSS的研究工作了，没有及时回复与维护，抱歉。您的问题的潜在原因可能是seco对patch进行tag赋予时导致的，那一块代码以及loss缺少对Nan的排除情况，所以当您跟我的环境以及设备不一致时，有可能导致训练崩溃。您可以首先在训练中期加入所设计LIG以及LIL，等到网络训练中期得到稳定的mask后再进行加入。

希望能对您有所帮助。

**Comment by wxqlab** [comment]
@Jingfeng-Tang 非常感谢您的回复

**Comment by wxqlab** [comment]
@zwyang6 您好，您的意思是考虑得到相对稳定的refined_pseudo_label再引入seg loss和reg loss继续训练model吗？因为如果在训练前期就引入seg loss和reg loss的话，由于refined_pseudo_label的细微差异（如：某个pixel的label从1变成0），从而导致seg loss在多次训练中的数值上存在差异，进而导致在多次训练中得不到一样的性能，对吗？

**Comment by zwyang6** [comment]
> [@zwyang6](https://github.com/zwyang6) 您好，您的意思是考虑得到相对稳定的refined_pseudo_label再引入seg loss和reg loss继续训练model吗？因为如果在训练前期就引入seg loss和reg loss的话，由于refined_pseudo_label的细微差异（如：某个pixel的label从1变成0），从而导致seg loss在多次训练中的数值上存在差异，进而导致在多次训练中得不到一样的性能，对吗？

您的情况我觉得主要是Loss_LIG以及Loss_LIL的在进行对比学习是训练崩溃导致的，一个潜在原因是mask的不稳定导致tag的分配错误，进而使得loss出现nan的情况，不知道你有没有遇到。您可以试一下在训练中期加入这两个loss试试看，先排除一下具体原因。
